torch/distributed/_tensor/experimental/attention.py

```diff
@@ -0,0 +1,467 @@
+                grad_key,
+                grad_value,
+            ) = torch.ops.aten._scaled_dot_product_flash_attention_backward(
+                grad_out=chunk_grad,
+                query=query,
+                key=key,
+                value=value,
+                out=output,
+                logsumexp=logsumexp,
+                cum_seq_q=cum_seq_q,
+                cum_seq_k=cum_seq_k,
+                max_q=max_q,
+                max_k=max_k,
+                dropout_p=dropout_p,
+                is_causal=is_causal_behavior.value,
+                philox_seed=philox_seed,
+                philox_offset=philox_offset,
+                scale=scale,
+            )
+        else:
+            grad_query = torch.zeros_like(query)
+            grad_key = torch.zeros_like(key)
+            grad_value = torch.zeros_like(value)
+
+        # TODO overlap grad communication
+        if i == 0:
+            out_grad_queries.append(grad_query)
+            out_grad_keys.append(grad_key)
+            out_grad_values.append(grad_value)
+        elif i > 0:
+            grad_dsts = [(-i) % size for i in range(size)]
+
+            grad_kv = torch.cat([grad_key.flatten(), grad_value.flatten()])
+            grad_kv = ft_c.permute_tensor(grad_kv, grad_dsts, pg)
+            grad_kv = ft_c.wait_tensor(grad_kv)
+            grad_key = grad_kv[: grad_key.numel()].reshape(grad_key.shape)
+            grad_value = grad_kv[grad_key.numel() :].reshape(grad_value.shape)
+
+            out_grad_queries.append(grad_query)
+            out_grad_keys.append(grad_key)
+            out_grad_values.append(grad_value)
+
+    # stack and sum to avoid accumulation errors
+    out_grad_query = torch.stack(out_grad_queries).sum(dim=0)
+    out_grad_key = torch.stack(out_grad_keys).sum(dim=0)
+    out_grad_value = torch.stack(out_grad_values).sum(dim=0)
+
+    return out_grad_query, out_grad_key, out_grad_value
+
+
+customized_ops = {
+    aten._scaled_dot_product_flash_attention.default: sdpa_handler,
+    aten._scaled_dot_product_flash_attention_backward.default: sdpa_backward_handler,
+}
+
+
+@contextlib.contextmanager
+def attention_context_parallel() -> Generator[None, None, None]:
+    """
+    This enables attention context parallel optimizations. Currently only supports ring attention for SDPA flash attention.
+    """
+    DTensor._op_dispatcher._custom_op_handlers.update(customized_ops)
+
+    yield
+
+    for custom_op in customized_ops:
+        DTensor._op_dispatcher._custom_op_handlers.pop(custom_op)
+
+
+class ContextParallel(ParallelStyle):
+    """
+    Applies context parallel optimizations to the attention layer. This will
+    work for nn.MultiHeadedAttention.
+
+    Currently only supports ring attention for SDPA flash attention.
+
+    See attention_context_parallel for more details.
+    """
+
+    def _apply(self, module: nn.Module, device_mesh: DeviceMesh) -> nn.Module:
+        if not isinstance(module, nn.MultiheadAttention):
```

Comment:
Maybe we should remove this restriction, given that it would also work with any custom attention layer as long as `F.sdpa` is called?

Reply:
done